src/movegen.cpp

```diff
@@ -236,14 +236,10 @@ namespace {
         while (b)
             *moveList++ = make_move(ksq, pop_lsb(&b));
 
-        if (Type != CAPTURES && pos.can_castle(CastlingRight(OO | OOO)))
-        {
-            if (!pos.castling_impeded(OO) && pos.can_castle(OO))
-                *moveList++ = make<CASTLING>(ksq, pos.castling_rook_square(OO));
-
-            if (!pos.castling_impeded(OOO) && pos.can_castle(OOO))
-                *moveList++ = make<CASTLING>(ksq, pos.castling_rook_square(OOO));
-        }
+        if (Type != CAPTURES)
+            for (CastlingRight cr : {OO, OOO})
+                if (!pos.castling_impeded(cr) && pos.can_castle(cr))
```

Comment:
Since `can_castle(cr)` is false in most positions, perhaps evaluate it before `!pos.castling_impeded(cr)`.  In fact if the king has moved it might be necessary to not evaluate `!pos.castling_impeded(cr)`.